kamu-core/src/infra/repos/dataset_impl.rs

```diff
@@ -110,60 +111,78 @@ where
 
         debug!(?current_head, ?last_seen, "Updating dataset summary");
 
+        let increment = self
+            .compute_summary_increment(&current_head, last_seen)
+            .await?;
+
+        let blank_summary: DatasetSummary = Self::prepare_blank_summary_for_update(&current_head);
+        let base_summary = if increment.seen_empty_prev_block_hash {
+            &blank_summary
+        } else {
+            prev.as_ref().unwrap_or(&blank_summary)
+        };
+
+        let summary =
+            UpdateSummaryIncrement::apply_increment_to(increment, &base_summary, &current_head);
+
+        self.write_summary(&summary).await?;
+
+        Ok(Some(summary))
+    }
+
+    async fn compute_summary_increment(
+        &self,
+        current_head: &Multihash,
+        last_seen: Option<&Multihash>,
+    ) -> Result<UpdateSummaryIncrement, GetSummaryError> {
+        let mut block_stream =
+            self.metadata_chain
+                .iter_blocks_interval(&current_head, last_seen, true);
+
         use tokio_stream::StreamExt;
-        let blocks: Vec<_> = self
-            .metadata_chain
-            .iter_blocks_interval(&current_head, last_seen)
-            .collect::<Result<_, _>>()
-            .await
-            .int_err()?;
-
-        let mut summary = prev.unwrap_or(DatasetSummary {
-            id: DatasetID::from_pub_key_ed25519(b""), // Will be replaced
-            kind: DatasetKind::Root,
-            last_block_hash: current_head.clone(),
-            dependencies: Vec::new(),
-            last_pulled: None,
-            num_records: 0,
-            data_size: 0,
-            checkpoints_size: 0,
-        });
+        let mut increment: UpdateSummaryIncrement = UpdateSummaryIncrement::new();
 
-        for (hash, block) in blocks.into_iter().rev() {
-            summary.last_block_hash = hash;
+        while let Some((_, block)) = block_stream.try_next().await.int_err()? {
+            increment.seen_empty_prev_block_hash = block.prev_block_hash.is_none();
 
             match block.event {
                 MetadataEvent::Seed(seed) => {
-                    summary.id = seed.dataset_id;
-                    summary.kind = seed.dataset_kind;
+                    increment.seen_id = increment.seen_id.or_else(|| Some(seed.dataset_id));
```

Comment:
A slightly more concise way to write it is:
```rust
increment.seen_id.get_or_insert(seed.dataset_id);
```
which might be a little less work for compiler to inline than a lambda function.

Reply:
Interesting